src/lib/stable/canister_methods/index.ts

```diff
@@ -284,59 +355,35 @@ function isDecoratorOverloadedWithoutParams<
  * @throws {Error} If methods have mismatched parameters or if invalid number of methods
  */
 function verifyInitAndPostUpgradeHaveTheSameParams(
-    idlTypes: IDL.FuncClass[]
+    a: IDL.Type[],
+    b: IDL.Type[]
 ): void {
-    if (idlTypes.length === 0 || idlTypes.length === 1) {
-        return;
-    }
-
-    if (idlTypes.length > 2) {
-        throwTooManyMethodsError(idlTypes);
-    }
-
-    const [init, postUpgrade] = findInitAndPostUpgradeMethods(idlTypes);
-
-    if (init === undefined || postUpgrade === undefined) {
-        throw new Error('Both init and postUpgrade methods must be defined');
-    }
+    const aSignature = idlToString(IDL.Func(a, []));
+    const bSignature = idlToString(IDL.Func(b, []));
 
-    // Get rid of the annotations to facilitate comparison
-    const initFunc = IDL.Func(init.argTypes, init.retTypes);
-    const postUpgradeFunc = IDL.Func(
-        postUpgrade.argTypes,
-        postUpgrade.retTypes
-    );
-    const initSignature = getFunctionSignature(initFunc);
-    const postUpgradeSignature = getFunctionSignature(postUpgradeFunc);
-
-    if (initSignature !== postUpgradeSignature) {
+    if (aSignature !== bSignature) {
         throw new Error(
-            `'init' and 'postUpgrade' methods must have the same parameters.\nFound:\n- init: ${initSignature}\n- postUpgrade: ${postUpgradeSignature}`
+            `'init' and 'postUpgrade' methods must have the same parameters.\nFound:\n${aSignature}\n${bSignature}`
         );
     }
 }
 
-function findInitAndPostUpgradeMethods(
-    idlTypes: IDL.FuncClass[]
-): [IDL.FuncClass | undefined, IDL.FuncClass | undefined] {
-    const init = idlTypes.find((type) => type.annotations.includes('init'));
-
-    const postUpgrade = idlTypes.find((type) =>
-        type.annotations.includes('post_upgrade')
-    );
-
-    return [init, postUpgrade];
-}
-
-function getFunctionSignature(func: IDL.FuncClass): string {
-    const visitor = new DidVisitor();
-    const result = visitor.visitFunc(func, getDefaultVisitorData());
-    return toDidString(result);
-}
-
-function throwTooManyMethodsError(idlTypes: IDL.FuncClass[]): never {
-    const methodSignatures = idlTypes.map(getFunctionSignature);
-    throw new Error(
-        `More than two init and postUpgrade methods found:\n${methodSignatures.join('\n')}`
-    );
+/**
+ * @internal
+ *
+ * Throws an error if a method is already defined in the class.
+ *
+ * @param methodName - The name of the method
+ * @param isDefined - Whether the method is already defined
+ * @throws {Error} If the method is already defined
+ */
+function throwIfMethodAlreadyDefined(
+    methodName: string,
```

Comment:
Use the keys of the DefinedSystemMethods type for this type

Reply:
Done